core/formats/formats_burst_trie.cpp

```diff
@@ -175,11 +164,81 @@ using volatile_byte_ref = volatile_ref<byte_type>;
 
 using feature_map_t = std::vector<irs::type_info::type_id>;
 
+template<typename T>
+struct Node {
+  T* next = nullptr;
+};
+
+template<typename T>
+struct IntrusiveList {
+ public:
+  IntrusiveList& operator=(const IntrusiveList&) = delete;
+  IntrusiveList(const IntrusiveList&) = delete;
+
+  IntrusiveList() noexcept = default;
+
+  IntrusiveList(IntrusiveList&& other) noexcept
+    : tail_{std::exchange(other.tail_, nullptr)} {}
+
+  IntrusiveList& operator=(IntrusiveList&& other) noexcept {
+    std::swap(tail_, other.tail_);
+    return *this;
+  }
+
+  void Append(IntrusiveList&& rhs) noexcept {
+    IRS_ASSERT(this != &rhs);
+    if (rhs.tail_ == nullptr) {
+      return;
+    }
+    if (tail_ == nullptr) {
+      tail_ = rhs.tail_;
+      rhs.tail_ = nullptr;
+      return;
+    }
+    // h1->t1->h1 h2->t2->h2
+    auto* head = tail_->next;
+    // h1->t1->h2->t2->h2
+    tail_->next = rhs.tail_->next;
+    // h1->t1->h2->t2->h1
+    rhs.tail_->next = head;
+    // h1->**->h2->t2/t1->h1
+    tail_ = rhs.tail_;
+    // h1->**->h2->t1->h1
+    rhs.tail_ = nullptr;
+  }
+
+  void PushFront(T& front) noexcept {
+    IRS_ASSERT(front.next == &front);
```

Comment:
Why we need this requirement? Isn't it safer to require "nullptr" in `next` for any incoming node to ensure it does not belongs to any other list. And then here set the right `next` value. BTW this assert will not fire if we have other list with single node and will pass by error that node here.

Reply:
I know. Because it's classic way to write circular list